_posts/2020-05-25-kinesis-streams-part-1.md

```diff
@@ -0,0 +1,439 @@
+
+KPL is another level of abstraction over API calls. Though simplifying certain aspects of writing to a stream, in my opinion, KPL can add an often unnecessary **layer of complexity** and comes with its own limitations you should be aware of when using it. With SDK you also avoid having an extra moving part in your application (the KPL daemon). You have full control over how you write your data. As we have discussed, you also have a greater responsibility, but, in most cases, just the few simple guidelines we discussed will be sufficient to let you sleep at night.
+
+Moreover, unlike KPL, most AWS SDK considerations apply across many other AWS services, instead of being just _Kinesis Streams_ specific. Thus understanding those aspects is probably a good idea if you are planning to interact with more AWS services.
+
+Finally, while there are several pros and cons to consider and argue about when choosing between AWS SDK and KPL, there is one that is the big pink elephant in the room. While KCL does have support in several languages, KPL is **[only supported in Java](https://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-integration.html)**
+
+![kpl](/img/kinesis/kpl.png)
+
+
+## Bonus: Kinesis Data Generator
+
+There is one more way to write data to a stream I wanted to mention. It is often useful to simulate data being written to the stream, f.ex. to test consumer's behavior. To simplify this process, there is a tool called **Kinesis Data Generator (KDG)**. You can find it in [GitHub](https://github.com/awslabs/amazon-kinesis-data-generator) or use the hosted UI [here](http://amzn.to/datagen). There is also an extensive [blog post](https://aws.amazon.com/blogs/big-data/test-your-streaming-data-solution-with-the-new-amazon-kinesis-data-generator/) on using KDG.
+
+# Monitoring a stream
+
+Now when we have been talking about failures for quite some time (because everything fails, all the time, right? :)), it's time to talk about how you can detect failures and monitor your stream's performance. _CloudWatch_ has a [set of metrics](https://docs.aws.amazon.com/streams/latest/dev/monitoring-with-cloudwatch.html) related to _Kinesis Streams_.
+
+The so-called "**basic**" metrics are enabled by default and tell you about the stream in overall. You can also enable the "**enhanced**" metrics, which allow you to monitor the stream on the level of individual shards. The enhanced shard-level metrics can be very useful, especially when you suspect having an "overheated" shard, but they will cost you extra. The basic stream-level metrics are free.
+
+Below are the most important metrics to keep an eye on when writing to a Kinesis stream:
+
+- `IncomingRecords`: The **number** of records **successfully** written to the Kinesis stream. This includes records from both, `PutRecord` and `PutRecods` operations. In case you want to separate the two, each of them has its own metric as well: `PutRecord.Success` and `PutRecords.Records` respectively.
```

Comment:
Remove comma: This includes records from both `PutRecord` and `PutRecods` operations.